lekko_client/__init__.py

```diff
@@ -1,9 +1,89 @@
 """Lekko Python SDK Client"""
 
-__version__ = "0.1.4"
+from enum import Enum
+from typing import Any, Dict, Optional, Type
 
-LEKKO_API_URL = "prod.api.lekko.dev:443"
-LEKKO_SIDECAR_URL = "localhost:50051"
+from google.protobuf.message import Message as ProtoMessage
 
-from lekko_client.client import APIClient, SidecarClient  # noqa
+from lekko_client.clients import (
+    APIClient,
+    CachedBackendClient,
+    CachedGitClient,
+    Client,
+    SidecarClient,
+)
+from lekko_client.constants import LEKKO_API_URL, LEKKO_SIDECAR_URL  # noqa
 from lekko_client.exceptions import *  # noqa
+from lekko_client.stores import MemoryStore
+
+__version__ = "0.1.4"
+
+__client: Client
+
+
+class Mode(Enum):
+    API = 1
+    SIDECAR = 2
+    CACHED_SERVER = 3
+    CACHED_GIT = 4
+
+
+def initialize(
+    mode: Mode,
+    owner_name: str,
+    repo_name: str,
+    api_key: Optional[str] = None,
+    context: Optional[Dict[str, Any]] = None,
+    git_repo_path: Optional[str] = None,
+) -> Client:
+    global __client
```

Comment:
I don't see it being used, is it for testing? Do we really need this global variable?

Reply:
i'm open to suggestions on refining the interface, but since we're caching and managing lifecycle, i wanted to make a static interface that has access to the intitialized client. the methods below all make use of `__client` when you call `lekko_client.get_bool` for example. i return the `client` here as a convenience but i don't think it will make sense for sdk users to hold onto it - better to call `lekko_client.initialize()` in their startup function and then just `lekko_client.get_*` everywhere else.

Reply:
Agree that it will be more convenient for some users. But I have a few concerns with this approach:
- `lekko_client.get_*` relies on user calling `lekko_client.initialize()` beforehand, and it can cause surprise runtime errors after refactoring or reusing a library in some other place. So at least we need to handle it with actionable message.
- it's not compatible with multi-repo setup and lead to unexpected behavior if `lekko.initialize()` is called multiple times. We probably need to have a lock here and throw actionable error if `initialize` is called more than once
- it pushed users to code style that is less composable (how to mix configs form different repo / different modes in one binary) and harder to test (we probably need to provide a helper to mock global client properly anyway)

I'm not against exposing this kind of interface as an option, but I think that we need to handle corner cases that I mentioned above (there are probably more) and clearly document limitations of this approach for users.
I can check with rest of the team if anyone has more opions.

Reply:
I think personally it makes more sense to allow users to initialize multiple times and users can manage their own global instance of a client if they want. Leaving the instance management to the users is also probably better for supporting patterns such as dependency injection which they may have

Reply:
my primary concern is that managing their own global instance can be a bit awkward and is susceptible to users inadvertently creating ephemeral instances - which is quite bad considering the two new modes are spinning up background threads to monitor for config updates.

there are ways to manage the instance for them and still make this a bit more DI friendly.

Reply:
i don't think we need to solve the multi-repo / multi-mode setup right now. we do need to manage the lifecycle properly and allow users to initialize + destroy the client. 

initialization would involve registration and filling the cache, and
destroy / close would involve deregistration.

i'm not a fan of global variables in general but i think whether or not the underlying implementation uses a global variable should be transparent to the user, as long as the edge cases @passichenko mentioned are handled.

we could expose a dedicated method to initialize the client instead of relying on the class constructor or `initialize()` method - e.g. 

```py
def cached_api_client(...)
...
```

which would invoke the constructor and perform any initialization needed. that way we can avoid issues around having the caller call `initialize()` twice. 

lmk if i'm missing any context on this.

Reply:
the only issue with the `cached_api_client()` is how cumbersome it would be for all api consumers to pass all initialization parameters just in case the client hasn't been constructed yet. for proper lifecycle management we really should encourage clients to call `initialize()` during their app startup - particularly so they don't incur the cost during request handling

Reply:
i guess the key philosophy here is dependency injection. when you have a non-static dependency that's used all across your codebase, what's the best practice for sharing that dependency in python?

in other languages (go, node, rust) we would inject the dependency vs having static global

Reply:
> what's the best practice for sharing that dependency in python?

I haven't really used python outside of relatively large projects inside a monorepo. In my experience we always end up going with dependency injection (usually manually passing things into constructors)

Reply:
Another thing that I don't understand with global instance pattern is how it helps with shutdown and deregistration. Who is responsible for properly closing lekko client?